### ISSUE TITLE
embedFileUtf8 prevents installation with cabal new-install problem with 

### ISSUE THREAD

**Issue by brittAnderson** [purescript/spago#702]
Was trying to install purescript and spago with cabal new style commands.  Purescript/purs worked without error, but spago fails. It loads and builds all the dependencies but fails with an error about the embedding of files and url fall backs. It can't find the file. I just took a guess and fixed it by putting the full path in the lines of Templates.hs like this, e.g.

`packagesDhall = $(embedFileUtf8 "/home/britt/progLang/purescript/spago/templates/packages.dhall")` where formerly there was only a `templates/packages.dhall` I had to do that for every such reference in this file. I suspect this has something to do with the 
`makeRelativeToProject` command in TH.hs file, e.g. 

` [| LT.decodeUtf8 $(makeRelativeToProject filePath >>= embedFile) |]`

If this relative file location could be fixed somehow than installation would only require having the right ghc versions, a couple of cabal new-install commands and some patience - at least on my machine. 

I guess this mistake does not come up with stack install for some reason. Don't know if it is worth fixing or not, but I thought I should report it since it doesn't seem that hard to fix manually if someone else runs into it. 
`

**Comment by f-f** [comment]
@brittAnderson thank you for reporting this! I am actually interested in switching the build to `cabal` entirely, so any instructions you have for building this are very welcome! (note: last time I used Cabal was more than 5y ago, so I am totally in the dark when it comes to the new build commands)

To get back to the issue at hand here - it looks like [`makeRelativeToProject`](https://www.stackage.org/haddock/lts-16.22/file-embed-0.0.11.2/src/Data.FileEmbed.html#makeRelativeToProject) is stack-specific. The link points to the implementation, but I'll report the docstring here for convenience:
```
-- | Take a relative file path and attach it to the root of the current
-- project.
--
-- The idea here is that, when building with Stack, the build will always be
-- executed with a current working directory of the root of the project (where
-- your .cabal file is located). However, if you load up multiple projects with
-- @stack ghci@, the working directory may be something else entirely.
--
-- This function looks at the source location of the Haskell file calling it,
-- finds the first parent directory with a .cabal file, and uses that as the
-- root directory for fixing the relative path.
```

So at this point I have two questions:
- can we patch this function somehow to work with both build systems?
- if not, is there a better way to include static files with Cabal?

**Comment by hdgarrood** [comment]
I don’t think a patch is needed, I think this is failing because the files we want to embed are not listed in extra-source-files. See https://github.com/haskell/cabal/issues/6118

**Comment by hdgarrood** [comment]
I think what happens is that newer versions of cabal build in a source distribution, ie they run `cabal sdist`, unpack the created tar.gz file into a temporary directory, and then build in that directory. This is nice because it means you can’t accidentally depend on files which you have forgotten to include in the source distribution.

**Comment by brittAnderson** [comment]
Contemporary cabal uses a nix inspired build system to create a store of packages that are hashed based on their versions and dependencies to allow multiple packages of the same name and even version to be stored on a single system without creating conflicts. The basic overview is here: [cabal.34.readthedocs](https://cabal.readthedocs.io/en/3.4/nix-local-build-overview.html). 

I believe that `cabal install` now defaults to the nix style, but I use `cabal new-install` out of habit. There are also `cabal new-build` and many other "new" variants of cabal commands. `cabal new-build` will build things deep in a sub-directory of the package while the new-install command gives you a link in your ~/.cabal/bin directory. 

So, to get purescript I just did `cabal new-install purescript` and it went to hackage and got me the current version and downloaded all the dependencies and spent awhile building all. For spago I cloned your repository and then cd'd into that directory to run `cabal new-install` alone which triggered cabal to look locally for the package cabal file and src. The only other step was using [ghcup](https://www.haskell.org/ghcup/) to get and set the versions of ghc (8.6.4 for purescript and 8.6.5 for spago) that seemed like the right ones for building. 

In practice it wasn't too complicated. Get/set ghc 8.6.4 install purescript. clone spago. Get/set 8.6.5 install spago. (and make sure I had nodejs). If it hadn't been for the path issue which @hdgarrod seems to have figured out it would only have meant waiting for things to compile. 

Hope this gives you the information you are looking for.
